File: crates/pray-core/src/literal.rs
use crate::{PrayError, PrayResult};
use serde::{Deserialize, Serialize};
use std::borrow::Cow;
use std::collections::BTreeMap;

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub enum LiteralValue {
    String(String),
    Symbol(String),
    Bool(bool),
    Null,
    Integer(i64),
    Array(Vec<LiteralValue>),
    Map(BTreeMap<String, LiteralValue>),
}

impl LiteralValue {
    pub fn as_string(&self) -> Option<&str> {
        match self {
            Self::String(value) | Self::Symbol(value) => Some(value),
            _ => None,
        }
    }

    pub fn as_bool(&self) -> Option<bool> {
        match self {
            Self::Bool(value) => Some(*value),
            _ => None,
        }
    }

    pub fn as_integer(&self) -> Option<i64> {
        match self {
            Self::Integer(value) => Some(*value),
            _ => None,
        }
    }

    pub fn as_array(&self) -> Option<&[LiteralValue]> {
        match self {
            Self::Array(values) => Some(values),
            _ => None,
        }
    }

    pub fn as_map(&self) -> Option<&BTreeMap<String, LiteralValue>> {
        match self {
            Self::Map(values) => Some(values),
            _ => None,
        }
    }
}

pub fn split_top_level(input: &str, separator: char) -> Vec<String> {
    let mut output = Vec::new();
    let mut start = 0usize;
    let mut depth = 0i32;
    let mut quote: Option<char> = None;
    let mut escaped = false;

    for (index, character) in input.char_indices() {
        if let Some(quote_char) = quote {
            if escaped {
                escaped = false;
            } else if character == '\\' {
                escaped = true;
            } else if character == quote_char {
                quote = None;
            }
            continue;
        }

        match character {
            '"' | '\'' => quote = Some(character),
            '[' | '{' | '(' => depth += 1,
            ']' | '}' | ')' => depth -= 1,
            _ if character == separator && depth == 0 => {
                output.push(input[start..index].trim().to_string());
                start = index + character.len_utf8();
            }
            _ => {}
        }
    }

    if start < input.len() {
        output.push(input[start..].trim().to_string());
    }

    output
        .into_iter()
        .filter(|segment| !segment.is_empty())
        .collect()
}

pub fn find_top_level(input: &str, token: &str) -> Option<usize> {
    let mut depth = 0i32;
    let mut quote: Option<char> = None;
    let mut escaped = false;
    let characters: Vec<(usize, char)> = input.char_indices().collect();

    let mut index = 0usize;
    while index < characters.len() {
        let (byte_index, character) = characters[index];
        if let Some(quote_char) = quote {
            if escaped {
                escaped = false;
            } else if character == '\\' {
                escaped = true;
            } else if character == quote_char {
                quote = None;
            }
            index += 1;
            continue;
        }

        match character {
            '"' | '\'' => quote = Some(character),
            '[' | '{' | '(' => depth += 1,
            ']' | '}' | ')' => depth -= 1,
            _ if depth == 0 && input[byte_index..].starts_with(token) => return Some(byte_index),
            _ => {}
        }
        index += 1;
    }
    None
}

pub fn is_balanced(input: &str) -> bool {
    let mut depth = 0i32;
    let mut quote: Option<char> = None;
    let mut escaped = false;

    for character in input.chars() {
        if let Some(quote_char) = quote {
            if escaped {
                escaped = false;
            } else if character == '\\' {
                escaped = true;
            } else if character == quote_char {
                quote = None;
            }
            continue;
        }

        match character {
            '"' | '\'' => quote = Some(character),
            '[' | '{' | '(' => depth += 1,
            ']' | '}' | ')' => depth -= 1,
            _ => {}
        }
    }

    depth == 0 && quote.is_none()
}

pub fn parse_literal(input: &str) -> PrayResult<LiteralValue> {
    let mut parser = LiteralParser::new(input);
    let value = parser.parse_value()?;
    parser.skip_whitespace();
    if !parser.is_finished() {
        return Err(PrayError::Parse {
            kind: "literal",
            message: format!("unexpected trailing input near {:?}", parser.remaining()),
        });
    }
    Ok(value)
}

pub fn parse_literal_map(input: &str) -> PrayResult<BTreeMap<String, LiteralValue>> {
    match parse_literal(input)? {
        LiteralValue::Map(entries) => Ok(entries),
        other => Err(PrayError::Parse {
            kind: "literal",
            message: format!("expected map literal, found {:?}", other),
        }),
    }
}

pub fn parse_literal_array(input: &str) -> PrayResult<Vec<LiteralValue>> {
    match parse_literal(input)? {
        LiteralValue::Array(entries) => Ok(entries),
        other => Err(PrayError::Parse {
            kind: "literal",
            message: format!("expected array literal, found {:?}", other),
        }),
    }
}

struct LiteralParser<'a> {
    input: &'a str,
    cursor: usize,
}

impl<'a> LiteralParser<'a> {
    fn new(input: &'a str) -> Self {
        Self { input, cursor: 0 }
    }

    fn is_finished(&self) -> bool {
        self.cursor >= self.input.len()
    }

    fn remaining(&self) -> &str {
        &self.input[self.cursor..]
    }

    fn skip_whitespace(&mut self) {
        while let Some(character) = self.peek() {
            if character.is_whitespace() {
                self.cursor += character.len_utf8();
            } else {
                break;
            }
        }
    }

    fn peek(&self) -> Option<char> {
        self.remaining().chars().next()
    }

    fn next(&mut self) -> Option<char> {
        let character = self.peek()?;
        self.cursor += character.len_utf8();
        Some(character)
    }

    fn parse_value(&mut self) -> PrayResult<LiteralValue> {
        self.skip_whitespace();
        match self.peek() {
            Some('"') | Some('\'') => self.parse_string(),
            Some(':') => self.parse_symbol(),
            Some('[') => self.parse_array(),
            Some('{') => self.parse_map(),
            Some(character) if character.is_ascii_digit() || character == '-' => {
                self.parse_integer_or_identifier()
            }
            Some(_) => self.parse_identifier(),
            None => Err(PrayError::Parse {
                kind: "literal",
                message: "unexpected end of input".to_string(),
            }),
        }
    }

    fn parse_string(&mut self) -> PrayResult<LiteralValue> {
        let quote = self.next().unwrap();
        let mut output = String::new();
        let mut escaped = false;
        while let Some(character) = self.next() {
            if escaped {
                output.push(match character {
                    'n' => '\n',
                    'r' => '\r',
                    't' => '\t',
                    '\\' => '\\',
                    '"' => '"',
                    '\'' => '\'',
                    other => other,
                });
                escaped = false;
                continue;
            }
            if character == '\\' {
                escaped = true;
                continue;
            }
            if character == quote {
                return Ok(LiteralValue::String(output));
            }
            output.push(character);
        }
        Err(PrayError::Parse {
            kind: "literal",
            message: "unterminated string literal".to_string(),
        })
    }

    fn parse_symbol(&mut self) -> PrayResult<LiteralValue> {
        self.next();
        let mut output = String::new();
        while let Some(character) = self.peek() {
            if character.is_alphanumeric() || matches!(character, '_' | '-' | '.' | '/') {
                output.push(character);
                self.next();
            } else {
                break;
            }
        }
        if output.is_empty() {
            return Err(PrayError::Parse {
                kind: "literal",
                message: "empty symbol".to_string(),
            });
        }
        Ok(LiteralValue::Symbol(output))
    }

    fn parse_array(&mut self) -> PrayResult<LiteralValue> {
        self.next();
        let mut values = Vec::new();
        loop {
            self.skip_whitespace();
            if matches!(self.peek(), Some(']')) {
                self.next();
                break;
            }
            values.push(self.parse_value()?);
            self.skip_whitespace();
            match self.peek() {
                Some(',') => {
                    self.next();
                }
                Some(']') => {
                    self.next();
                    break;
                }
                _ => {
                    return Err(PrayError::Parse {
                        kind: "literal",
                        message: "expected ',' or ']'".to_string(),
                    })
                }
            }
        }
        Ok(LiteralValue::Array(values))
    }

    fn parse_map(&mut self) -> PrayResult<LiteralValue> {
        self.next();
        let mut entries = BTreeMap::new();
        loop {
            self.skip_whitespace();
            if matches!(self.peek(), Some('}')) {
                self.next();
                break;
            }
            let key = self.parse_map_key()?;
            self.skip_whitespace();
            if self.remaining().starts_with("=>") {
                self.next();
                self.next();
            } else if matches!(self.peek(), Some(':')) {
                self.next();
            } else {
                return Err(PrayError::Parse {
                    kind: "literal",
                    message: "expected ':' or '=>' after map key".to_string(),
                });
            }
            let value = self.parse_value()?;
            entries.insert(key, value);
            self.skip_whitespace();
            match self.peek() {
                Some(',') => {
                    self.next();
                }
                Some('}') => {
                    self.next();
                    break;
                }
                _ => {
                    return Err(PrayError::Parse {
                        kind: "literal",
                        message: "expected ',' or '}'".to_string(),
                    })
                }
            }
        }
        Ok(LiteralValue::Map(entries))
    }

    fn parse_map_key(&mut self) -> PrayResult<String> {
        self.skip_whitespace();
        match self.peek() {
            Some('"') | Some('\'') => match self.parse_string()? {
                LiteralValue::String(value) => Ok(value),
                _ => unreachable!(),
            },
            Some(':') => match self.parse_symbol()? {
                LiteralValue::Symbol(value) => Ok(value),
                _ => unreachable!(),
            },
            Some(character) if is_identifier_start(character) => self.parse_identifier_name(),
            _ => Err(PrayError::Parse {
                kind: "literal",
                message: "invalid map key".to_string(),
            }),
        }
    }

    fn parse_integer_or_identifier(&mut self) -> PrayResult<LiteralValue> {
        let start = self.cursor;
        if matches!(self.peek(), Some('-')) {
            self.next();
        }
        while matches!(self.peek(), Some(character) if character.is_ascii_digit() || character == '_')
        {
            self.next();
        }
        if matches!(self.peek(), Some('.')) {
            return self.parse_identifier_from(start);
        }
        let text = self.input[start..self.cursor].replace('_', "");
        let parsed = text.parse::<i64>().map_err(|error| PrayError::Parse {
            kind: "literal",
            message: error.to_string(),
        })?;
        Ok(LiteralValue::Integer(parsed))
    }

    fn parse_identifier(&mut self) -> PrayResult<LiteralValue> {
        let identifier = self.parse_identifier_name()?;
        match identifier.as_str() {
            "true" => Ok(LiteralValue::Bool(true)),
            "false" => Ok(LiteralValue::Bool(false)),
            "nil" => Ok(LiteralValue::Null),
            other => Ok(LiteralValue::String(other.to_string())),
        }
    }

    fn parse_identifier_name(&mut self) -> PrayResult<String> {
        let start = self.cursor;
        if !matches!(self.peek(), Some(character) if is_identifier_start(character)) {
            return Err(PrayError::Parse {
                kind: "literal",
                message: "expected identifier".to_string(),
            });
        }
        self.next();
        while matches!(self.peek(), Some(character) if is_identifier_continue(character)) {
            self.next();
        }
        Ok(self.input[start..self.cursor].to_string())
    }

    fn parse_identifier_from(&mut self, start: usize) -> PrayResult<LiteralValue> {
        while matches!(self.peek(), Some(character) if is_identifier_continue(character) || character == '.')
        {
            self.next();
        }
        Ok(LiteralValue::String(
            self.input[start..self.cursor].to_string(),
        ))
    }
}

fn is_identifier_start(character: char) -> bool {
    character.is_ascii_alphabetic() || character == '_'
}

fn is_identifier_continue(character: char) -> bool {
    character.is_ascii_alphanumeric() || matches!(character, '_' | '-' | '.' | '/')
}

pub fn prepare_parser_lines(text: &str) -> Vec<Cow<'_, str>> {
    text.lines().map(prepare_parser_line).collect()
}

fn prepare_parser_line(line: &str) -> Cow<'_, str> {
    trim_end_cow(strip_line_comment(line))
}

pub fn strip_line_comment(line: &str) -> Cow<'_, str> {
    let mut quote: Option<char> = None;
    let mut escaped = false;
    for (index, character) in line.char_indices() {
        if let Some(quote_char) = quote {
            if escaped {
                escaped = false;
            } else if character == '\\' {
                escaped = true;
            } else if character == quote_char {
                quote = None;
            }
            continue;
        }
        match character {
            '"' | '\'' => quote = Some(character),
            '#' => return Cow::Borrowed(&line[..index]),
            _ => {}
        }
    }
    Cow::Borrowed(line)
}

fn trim_end_cow<'a>(value: Cow<'a, str>) -> Cow<'a, str> {
    match value {
        Cow::Borrowed(slice) => {
            let trimmed = slice.trim_end();
            if trimmed.len() == slice.len() {
                Cow::Borrowed(slice)
            } else {
                Cow::Owned(trimmed.to_string())
            }
        }
        Cow::Owned(mut owned) => {
            let trimmed_len = owned.trim_end().len();
            owned.truncate(trimmed_len);
            Cow::Owned(owned)
        }
    }
}

File: rubygems/pray-cli/lib/pray/literal.rb
# frozen_string_literal: true

module Pray
  module Literal
    LiteralValue = Struct.new(:kind, :value, keyword_init: true) do
      def string? = kind == :string || kind == :symbol
      def as_string = string? ? value : nil
      def as_bool = kind == :bool ? value : nil
      def as_integer = kind == :integer ? value : nil
      def as_array = kind == :array ? value : nil
      def as_map = kind == :map ? value : nil
    end

    module_function

    def split_top_level(input, separator)
      output = []
      start = 0
      depth = 0
      quote = nil
      escaped = false

      input.each_char.with_index do |character, index|
        if quote
          if escaped
            escaped = false
          elsif character == "\\"
            escaped = true
          elsif character == quote
            quote = nil
          end
          next
        end

        case character
        when '"', "'"
          quote = character
        when "[", "{", "("
          depth += 1
        when "]", "}", ")"
          depth -= 1
        else
          if character == separator && depth.zero?
            segment = input[start...index].strip
            output << segment unless segment.empty?
            start = index + 1
          end
        end
      end

      tail = input[start..].strip
      output << tail unless tail.empty?
      output
    end

    def find_top_level(input, token)
      depth = 0
      quote = nil
      escaped = false
      index = 0

      while index < input.length
        character = input[index]
        if quote
          if escaped
            escaped = false
          elsif character == "\\"
            escaped = true
          elsif character == quote
            quote = nil
          end
          index += 1
          next
        end

        case character
        when '"', "'"
          quote = character
        when "[", "{", "("
          depth += 1
        when "]", "}", ")"
          depth -= 1
        else
          return index if depth.zero? && input[index..].start_with?(token)
        end
        index += 1
      end
      nil
    end

    def is_balanced?(input)
      depth = 0
      quote = nil
      escaped = false

      input.each_char do |character|
        if quote
          if escaped
            escaped = false
          elsif character == "\\"
            escaped = true
          elsif character == quote
            quote = nil
          end
          next
        end

        case character
        when '"', "'"
          quote = character
        when "[", "{", "("
          depth += 1
        when "]", "}", ")"
          depth -= 1
        end
      end

      depth.zero? && quote.nil?
    end

    def parse_literal(input)
      parser = Parser.new(input)
      value = parser.parse_value
      parser.skip_whitespace
      unless parser.finished?
        raise Error.parse("literal", "unexpected trailing input near #{parser.remaining.inspect}")
      end
      value
    end

    def parse_literal_map(input)
      value = parse_literal(input)
      raise Error.parse("literal", "expected map literal, found #{value.inspect}") unless value.kind == :map

      value.value
    end

    def parse_literal_array(input)
      value = parse_literal(input)
      raise Error.parse("literal", "expected array literal, found #{value.inspect}") unless value.kind == :array

      value.value
    end

    def prepare_parser_lines(text)
      text.lines.map { |line| prepare_parser_line(line) }
    end

    def prepare_parser_line(line)
      strip_line_comment(line).rstrip
    end

    def strip_line_comment(line)
      quote = nil
      escaped = false
      line.each_char.with_index do |character, index|
        if quote
          if escaped
            escaped = false
          elsif character == "\\"
            escaped = true
          elsif character == quote
            quote = nil
          end
          next
        end

        case character
        when '"', "'"
          quote = character
        when "#"
          return line[0...index]
        end
      end
      line
    end

    class Parser
      def initialize(input)
        @input = input
        @cursor = 0
      end

      def finished? = @cursor >= @input.length
      def remaining = @input[@cursor..]

      def skip_whitespace
        while (character = peek) && character.match?(/\s/)
          @cursor += 1
        end
      end

      def peek = remaining[0]

      def next_character
        character = peek
        @cursor += 1 if character
        character
      end

      def parse_value
        skip_whitespace
        case peek
        when '"', "'"
          parse_string
        when ":"
          parse_symbol
        when "["
          parse_array
        when "{"
          parse_map
        when "-", "0".."9"
          parse_integer_or_identifier
        when nil
          raise Error.parse("literal", "unexpected end of input")
        else
          parse_identifier
        end
      end

      def parse_string
        quote = next_character
        output = +""
        escaped = false
        while (character = next_character)
          if escaped
            output << case character
                      when "n" then "\n"
                      when "r" then "\r"
                      when "t" then "\t"
                      when "\\" then "\\"
                      when "\"", "'" then character
                      else character
                      end
            escaped = false
            next
          end
          if character == "\\"
            escaped = true
            next
          end
          if character == quote
            return LiteralValue.new(kind: :string, value: output)
          end
          output << character
        end
        raise Error.parse("literal", "unterminated string literal")
      end

      def parse_symbol
        next_character
        output = +""
        while (character = peek) && character.match?(%r{[[:alnum:]_\-.\\/]})
          output << character
          next_character
        end
        raise Error.parse("literal", "empty symbol") if output.empty?

        LiteralValue.new(kind: :symbol, value: output)
      end

      def parse_array
        next_character
        values = []
        loop do
          skip_whitespace
          break if peek == "]" && next_character

          values << parse_value
          skip_whitespace
          case peek
          when ","
            next_character
          when "]"
            next_character
            break
          else
            raise Error.parse("literal", "expected ',' or ']'")
          end
        end
        LiteralValue.new(kind: :array, value: values)
      end

      def parse_map
        next_character
        entries = {}
        loop do
          skip_whitespace
          break if peek == "}" && next_character

          key = parse_map_key
          skip_whitespace
          if remaining.start_with?("=>")
            @cursor += 2
          elsif peek == ":"
            next_character
          else
            raise Error.parse("literal", "expected ':' or '=>' after map key")
          end
          entries[key] = parse_value
          skip_whitespace
          case peek
          when ","
            next_character
          when "}"
            next_character
            break
          else
            raise Error.parse("literal", "expected ',' or '}'")
          end
        end
        LiteralValue.new(kind: :map, value: entries)
      end

      def parse_map_key
        skip_whitespace
        case peek
        when '"', "'"
          value = parse_string
          value.value
        when ":"
          parse_symbol.value
        when "a".."z", "A".."Z", "_"
          parse_identifier_name
        else
          raise Error.parse("literal", "invalid map key")
        end
      end

      def parse_integer_or_identifier
        start = @cursor
        next_character if peek == "-"
        while peek&.match?(/[0-9_]/)
          next_character
        end
        if peek == "."
          return parse_identifier_from(start)
        end
        text = @input[start...@cursor].delete("_")
        LiteralValue.new(kind: :integer, value: Integer(text))
      rescue ArgumentError => error
        raise Error.parse("literal", error.message)
      end

      def parse_identifier
        identifier = parse_identifier_name
        case identifier
        when "true" then LiteralValue.new(kind: :bool, value: true)
        when "false" then LiteralValue.new(kind: :bool, value: false)
        when "nil" then LiteralValue.new(kind: :null, value: nil)
        else LiteralValue.new(kind: :string, value: identifier)
        end
      end

      def parse_identifier_name
        raise Error.parse("literal", "expected identifier") unless identifier_start?(peek)

        start = @cursor
        next_character
        while peek && identifier_continue?(peek)
          next_character
        end
        @input[start...@cursor]
      end

      def parse_identifier_from(start)
        while peek && (identifier_continue?(peek) || peek == ".")
          next_character
        end
        LiteralValue.new(kind: :string, value: @input[start...@cursor])
      end

      def identifier_start?(character)
        character && (character.match?(/[[:alpha:]]/) || character == "_")
      end

      def identifier_continue?(character)
        character.match?(%r{[[:alnum:]_\-.\\/]})
      end
    end
  end
end
